layers/+lang/haskell/packages.el

```diff
@@ -68,123 +74,145 @@
+    (dolist (mode spacemacs--haskell-modes)
+      (spacemacs/declare-prefix-for-mode mode "md" "haskell/debug")
+      (spacemacs/declare-prefix-for-mode mode "mr" "haskell/refactor")
+      (spacemacs/declare-prefix-for-mode mode "mh" "haskell/documentation")
+      (spacemacs/set-leader-keys-for-major-mode mode
+        "gb" 'xref-pop-marker-stack
+        ;; FIXME: the following three collides with haskell-mode
+        "ht" 'dante-type-at
+        "hT" 'spacemacs-haskell/dante-insert-type
+        "hi" 'dante-info
+        "rs" 'dante-auto-fix
+        "se" 'dante-eval-block
+        "sr" 'dante-restart))))
 
 (defun haskell/init-attrap ()
   (use-package attrap
     :defer t))
 
-(defun haskell/init-helm-hoogle ()
-  (use-package helm-hoogle
-    :defer t
-    :init
-    (dolist (mode haskell-modes)
-      (spacemacs/set-leader-keys-for-major-mode mode "hf" 'helm-hoogle))))
+
+;; lsp backend
 
-(defun haskell/post-init-flycheck ()
-  (progn
-    (add-hook 'dante-mode-hook
-              (lambda () (flycheck-add-next-checker 'haskell-dante '(warning . haskell-hlint))))
-    (spacemacs/enable-flycheck 'haskell-mode))
-  )
+(defun haskell/init-lsp-haskell ()
+  (use-package lsp-haskell
+    :defer t
+    :custom
+    (lsp-haskell-hlint-on haskell-enable-hlint)))
 
-(defun haskell/init-flycheck-haskell ()
-  (use-package flycheck-haskell
-    :commands flycheck-haskell-configure
-    :init (add-hook 'flycheck-mode-hook 'flycheck-haskell-configure)))
+
+;; formatter
 
-(defun haskell/post-init-ggtags ()
-  (add-hook 'haskell-mode-local-vars-hook #'spacemacs/ggtags-mode-enable)
-  (add-hook 'haskell-literate-mode-local-vars-hook #'spacemacs/ggtags-mode-enable)
-  ;; renamed as of 04/2020, delete in due course
-  (add-hook 'literate-haskell-mode-local-vars-hook #'spacemacs/ggtags-mode-enable))
+(defun haskell/init-format-all ()
+  (use-package format-all
+    :defer t
+    :after haskell-mode
+    :config
+    (progn
+      ;; depending on the formatter, it may or may not support `haskell-literate-mode'
+      (let ((modes (pcase haskell-formatter-backend
+                     ('brittany '(haskell-mode literate-haskell-mode))
+                     ('hindent '(haskell-mode literate-haskell-mode))
+                     ('ormolu '(haskell-mode))
+                     ('stylish-haskell '(haskell-mode))
+                     (_))))
+       (spacemacs/add-to-hooks 'format-all-mode modes
+        (dolist (mode modes)
+          (spacemacs/set-leader-keys-for-major-mode mode
+            "==" 'format-all-buffer
+            "T=" 'format-all-mode))))
+      ;; also set lsp formattign provider
+      ;; NOTE: except of 'hindent which is not currently supported (April 2021)
+      (when (and (eq haskell-completion-backend 'lsp)
+                 (memq haskell-formatter-backend
+                       '(brittany ormolu stylish-haskell fourmolu floskell)))
+        (setq lsp-haskell-formatting-provider (symbol-name haskell-formatter-backend))))))
```

Comment:
Does this work properly, given that `lsp-haskell-formatting-provider` is a `defcustom`? Can't we just let people set the already-existing customization variable that controls this, rather than adding another layer of abstraction here?

Reply:
The thing is `hindent` doesn't work with `lsp-haskell`, so I need to add this test.
Also, there was an old variable `haskell-enable-hindent`, I don't want to cause a hard beak. So for compatibility ,`haskell-formatter-backend` would default to `hindent`.

Reply:
I mean, why not do:
- `haskell-enable-hindent`: enable `hindent`, unless the backend is `lsp`, in which case give an error.
- No other formatting configuration: `lsp` users can use the configuration variables provided by `lsp-haskell`

Reply:
Also, sorry I'm not that familiar with `format-all`, do we want it if we have the `lsp` support for formatting?

Reply:
@michaelpj 

The new behaviour would be:
- If `lsp` is not used, and `haskell-enable-hindent` is non-nil, use `hindent.
- If `lsp` is used, and `haskell-completion-backend` is non-nil, use its value.
- Otherwise, search for a supported formatter in the PATH.

Does this look good for you?

Reply:
> If lsp is used, and haskell-completion-backend is non-nil, use its value.

I think you mean `haskell-formatting-provider` there?

> Otherwise, search for a supported formatter in the PATH.

I'd prefer not to have a fallback like this, it's very implicit. I might have multiple formatters, or none, and in any case I'd like to use whatever I've set up with `lsp` instead of those!

I didn't quite understand what you were suggesting, are you suggesting to keep `haskell-formatting-provider` as a variable or not?

Reply:
I think this is an admirable attempt to make everything work. I guess I wonder if we could get away with trying less hard:

- Don't provide any formatting options when using `dante` (there are none currently!)
- Don't provide any formatting options when using `lsp` (the `lsp-haskell` options do the job)
- Remove `haskell-enable-hindent` (you can enable `hindent-mode` yourself if you still want it)

I'm biased, of course, since I hope the community will continue to converge on the LSP-based solutions! But it would be nice to actually benefit from that and just leave the bulk of the configuration to `lsp-mode` and `lsp-haskell`.